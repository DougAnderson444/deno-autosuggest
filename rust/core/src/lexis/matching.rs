use std::fmt;
use std::collections::HashSet;
use crate::damlev::DamerauLevenshtein;
use crate::jaccard::Jaccard;
use super::{Word, Text};


const LENGTH_THRESHOLD:  f64 = 0.51;
const JACCARD_THRESHOLD: f64 = 0.41;
const DAMLEV_THRESHOLD:  f64 = 0.21;

thread_local! {
    static DAMLEV:  DamerauLevenshtein<char> = DamerauLevenshtein::new();
    static JACCARD: Jaccard<char>            = Jaccard::new();
}


#[derive(Clone, PartialEq)]
pub struct WordMatch {
    pub query:  MatchSide,
    pub record: MatchSide,
    pub typos:  usize,
    pub fin:    bool,
}


#[derive(Clone, PartialEq, Debug)]
pub struct MatchSide {
    pub ix:      usize,
    pub len:     usize,
    pub slice:   (usize, usize),
    pub primary: bool,
}


impl MatchSide {
    pub fn new(len: usize, slice: (usize, usize), primary: bool) -> Self {
        MatchSide { ix: 0, len, slice, primary }
    }
}


impl fmt::Debug for WordMatch {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "WordMatch {{ ")?;

        for i in 0 .. self.record.len {
            if i == self.record.slice.0 { write!(f, "[")?; }
            write!(f, "r")?;
            if i + 1 == self.record.slice.1 { write!(f, "]")?; }
        }

        write!(f, " /{}/ ", self.typos)?;

        for i in 0 .. self.query.len {
            if i == self.query.slice.0 { write!(f, "[")?; }
            write!(f, "q")?;
            if i + 1 == self.query.slice.1 { write!(f, "]")?; }
        }
        if !self.fin {
            write!(f, "..")?;
        }

        write!(f, " }}")?;
        Ok(())
    }
}


pub fn text_match(rtext: &Text<&[char]>, qtext: &Text<&[char]>) -> Vec<WordMatch> {
    let capacity = min!(rtext.words.len(), qtext.words.len());
    let mut taken: HashSet<usize> = HashSet::with_capacity(capacity);
    let mut matches: Vec<WordMatch> = Vec::with_capacity(capacity);

    for (i, qword) in qtext.words.iter().enumerate() {
        let mut found: Option<WordMatch> = None;

        for (j, rword) in rtext.words.iter().enumerate() {
            if taken.contains(&j) { continue; }
            if let Some(mut m) = word_match(rword, qword, &rtext.chars, &qtext.chars) {
                m.query.ix  = i;
                m.record.ix = j;
                if found.is_none() && !m.record.primary {
                    found = Some(m);
                    continue;
                }
                if m.record.primary {
                    found = Some(m);
                    break;
                }
            }
        }

        if let Some(m) = found {
            taken.insert(m.record.ix);
            matches.push(m);
        }
    }

    matches
}


pub fn word_match(rword: &Word, qword: &Word, rchars: &[char], qchars: &[char]) -> Option<WordMatch> {
    if qword.is_empty() || rword.is_empty() {
        return None;
    }
    if !length_check(rword, qword) {
        return None;
    }
    if !jaccard_check(rword, qword, rchars, qchars) {
        return None;
    }

    let rchars = rword.view(rchars);
    let qchars = qword.view(qchars);
    let mut best_match = None;

    DAMLEV.with(|damlev| {
        damlev.distance(qchars, rchars);
        let dists  = &*damlev.dists.borrow();

        let left  = if qword.fin { max!(qword.stem, rword.stem) } else { qword.stem } - 1;
        let right = max!(qword.len(), rword.len()) + 1;

        if right <= left { return; }

        let range = (left .. right).rev();

        for rlen in range.clone() {
            for qlen in range.clone() {
                // Out of bounds.
                if qlen > qword.len() { continue; }
                if rlen > rword.len() { continue; }
                // Left margin is for insertion/deletion, not for both prefixes at the same time.
                if rlen == left  && qlen == left  { continue; }
                // Compare full words only if query is finished.
                if qword.fin && rlen < rword.stem { break; }
                if qword.fin && qlen < qword.stem { break; }
                // Words with 2+ insertions/deletions are mismatched by default.
                if (qlen as isize - rlen as isize).abs() > 1 { continue; }

                let dist = dists.get(qlen + 1, rlen + 1);
                let rel  = dist as f64 / max!(qlen, rlen, 1) as f64;

                if rel > DAMLEV_THRESHOLD { continue; }

                match best_match {
                    None => {
                        best_match = Some(WordMatch {
                            query:  MatchSide::new(qlen,        (0, qlen), qword.is_primary()),
                            record: MatchSide::new(rword.len(), (0, rlen), rword.is_primary()),
                            typos:  dist,
                            fin:    qword.fin || rword.len() == rlen,
                        });
                    },
                    Some(ref mut m) => {
                        if m.typos <= dist { continue; }
                        m.record.slice = (0, rlen);
                        m.typos        = dist;
                        m.fin          = qword.fin || rword.len() == rlen;
                    },
                }

                if dist == 0 {
                    break;
                }
            }
        }
    });

    best_match
}


pub fn length_check(rword: &Word, qword: &Word) -> bool {
    let qlen  = qword.len();
    let rlen  = if qword.fin { rword.len() } else { min!(qlen, rword.len()) };

    if qlen <= 3 || rlen <= 3 {
        return qlen == rlen;
    }

    let long  = max!(qlen, rlen);
    let short = min!(qlen, rlen);
    let dist  = 1.0 - (short as f64 / long as f64);

    dist < LENGTH_THRESHOLD
}


pub fn jaccard_check(rword: &Word, qword: &Word, rchars: &[char], qchars: &[char]) -> bool {
    let rchars = rword.view(rchars);
    let qchars = qword.view(qchars);
    let rslice = if qword.fin { &rchars } else { &rchars[.. min!(qword.len(), rword.len())] };
    let dist   = JACCARD.with(|j| j.rel_dist(&rslice, &qchars));
    dist < JACCARD_THRESHOLD
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::lexis::Chars;
    use crate::lang::lang_english;
    use super::{Word, Text, word_match, text_match};


    fn text(s: &str) -> Text<Vec<char>> {
        Text::from_str(s).split(&Chars::Whitespaces)
    }


    // Match word
    // ----------------------------------------------------------------

    #[test]
    fn match_word_empty_both() {
        let qchars = "".chars().collect::<Vec<_>>();
        let rchars = "".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_eq!(word_match(&rword, &qword, &rchars[..], &qchars[..]), None);
    }


    #[test]
    fn match_word_empty_record() {
        let qchars = "mailbox".chars().collect::<Vec<_>>();
        let rchars = "".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_eq!(word_match(&rword, &qword, &rchars[..], &qchars[..]), None);
    }


    #[test]
    fn match_word_empty_query() {
        let qchars = "".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_eq!(word_match(&rword, &qword, &rchars[..], &qchars[..]), None);
    }


    // Match word: prefixed by same length word
    // ----------------------------------------------------------------

    #[test]
    fn match_word_full_strict() {
        let qchars = "mailbox".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_full_fuzzy_insertion() {
        let qchars = "mailybox".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_full_fuzzy_deletion() {
        let qchars = "mailox".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_full_fuzzy_transposition() {
        let qchars = "maiblox".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_full_query_too_long() {
        let qchars1 = "mailboxes".chars().collect::<Vec<_>>();
        let qchars2 = "mailboxes".chars().collect::<Vec<_>>();
        let rchars  = "mail"     .chars().collect::<Vec<_>>();
        let qword1  = Word::new(qchars1.len()).fin(true);
        let qword2  = Word::new(qchars2.len()).fin(false);
        let rword   = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword1, &rchars[..], &qchars1[..]));
        assert_debug_snapshot!(word_match(&rword, &qword2, &rchars[..], &qchars2[..]));
    }

    #[test]
    fn match_word_full_stem() {
        let     rchars  = "universe"  .chars().collect::<Vec<_>>();
        let     qchars1 = "university".chars().collect::<Vec<_>>();
        let     qchars2 = "university".chars().collect::<Vec<_>>();
        let mut rword   = Word::new(rchars.len());
        let mut qword1  = Word::new(qchars1.len());
        let     qword2  = Word::new(qchars2.len());

        let lang = lang_english();
        qword1.stem(&qchars1[..], &lang);
        rword .stem(&rchars[..],  &lang);

        assert_debug_snapshot!(word_match(&rword, &qword1, &rchars[..], &qchars1[..]));
        assert_debug_snapshot!(word_match(&rword, &qword2, &rchars[..], &qchars2[..]));
    }


    // Match word: prefixed by lesser length word
    // ----------------------------------------------------------------

    #[test]
    fn match_word_partial_strict() {
        let qchars = "mailb".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword = Word::new(qchars.len()).fin(false);
        let rword = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_partial_fuzzy_insertion() {
        let qchars = "maiylb".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_partial_fuzzy_deletion() {
        let qchars = "maib".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_partial_fuzzy_transposition() {
        let qchars = "malib".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    // Match text
    // ----------------------------------------------------------------

    #[test]
    fn match_text_empty_both() {
        let qtext = text("").fin(false);
        let rtext = text("");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_empty_one() {
        let qtext = text("mailbox").fin(false);
        let rtext = text("");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
        assert_debug_snapshot!(text_match(&qtext.to_ref(), &rtext.to_ref()));
    }


    #[test]
    fn match_text_singleton_equality() {
        let qtext = text("mailbox").fin(false);
        let rtext = text("mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_singleton_typos() {
        let qtext = text("maiblox").fin(false);
        let rtext = text("mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_pair_first() {
        let qtext = text("yelow").fin(false);
        let rtext = text("yellow mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_pair_second() {
        let qtext = text("maiblox").fin(false);
        let rtext = text("yellow mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_pair_unfinished() {
        let qtext = text("maiblox yel").fin(false);
        let rtext = text("yellow mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_intersection() {
        let qtext = text("big malibox yelo").fin(false);
        let rtext = text("small yellow metal mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_best_rword() {
        let mut qtext = text("the").fin(false);
        let mut rtext = text("the theme");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));

        let lang = lang_english();
        qtext = qtext.mark_pos(&lang);
        rtext = rtext.mark_pos(&lang);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_regression_best_match() {
        let qtext = text("sneak").fin(false);
        let rtext = text("sneaky");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }
}