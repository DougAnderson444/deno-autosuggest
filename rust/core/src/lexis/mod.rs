mod pattern;
mod matching;
mod tokenizing;

use std::fmt;
use std::borrow::Cow;
pub use matching::{WordMatch, MatchSide};
pub use pattern::{Chars, CharPattern};


pub fn tokenize_query<'a>(source: &'a [char]) -> Text<Cow<'a, [char]>> {
    Text::new_cow(Cow::Borrowed(source))
        .fin(false)
        .split(&[Chars::Whitespaces, Chars::Control, Chars::Punctuation])
        .strip(&[Chars::NotAlphaNum])
        .lower()
}


pub fn tokenize_record<'a>(source: &'a [char]) -> Text<Cow<'a, [char]>> {
    Text::new_cow(Cow::Borrowed(source))
        .split(&[Chars::Whitespaces, Chars::Control])
        .strip(&[Chars::NotAlphaNum])
        .lower()
}


#[derive(Clone, PartialEq)]
pub struct Text<T: AsRef<[char]>> {
    pub source: T,
    pub words: Vec<Word<T>>,
}


impl<'a> Text<Cow<'a, [char]>> {
    pub fn new_cow(source: Cow<'a, [char]>) -> Text<Cow<'a, [char]>> {
        let cloned = match source {
            Cow::Borrowed(s) => Cow::Borrowed(&s[..]),
            Cow::Owned(ref s) => Cow::Owned(s[..].to_vec()),
        };
        Text {
            source,
            words: vec![Word::new_cow(cloned)],
        }
    }
}


impl Text<Vec<char>> {
    pub fn new_owned(source: Vec<char>) -> Text<Vec<char>> {
        let cloned = source.clone();
        Text {
            source: source,
            words: vec![Word::new_owned(cloned)],
        }
    }
}


impl<T: AsRef<[char]>> Text<T> {
    pub fn fin(mut self, fin: bool) -> Self {
        if let Some(word) = self.words.last_mut() {
            word.fin = fin;
        }
        self
    }

    pub fn is_empty(&self) -> bool {
        self.words.is_empty()
    }

    pub fn to_owned(&self) -> Text<Vec<char>> {
        Text {
            source: self.source.as_ref().to_vec(),
            words:  self.words.iter().map(|w| w.to_owned()).collect(),
        }
    }

    pub fn to_cow<'a>(&'a self) -> Text<Cow<'a, [char]>> {
        Text {
            source: Cow::Borrowed(self.source.as_ref()),
            words:  self.words.iter().map(|w| w.to_cow()).collect(),
        }
    }
}


impl<T: AsRef<[char]>> fmt::Debug for Text<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Text {{")?;
        for word in &self.words {
            write!(f, " \"")?;
            for ch in word.chars.as_ref().iter() {
                write!(f, "{}", ch)?;
            }
            write!(f, "\"")?;
            if !word.fin { write!(f, "..")?; }
        }
        write!(f, " }}")?;
        Ok(())
    }
}


#[derive(Clone, PartialEq)]
pub struct Word<T: AsRef<[char]>> {
    pub source: T,
    pub slice:  (usize, usize),
    pub chars:  T,
    pub fin:    bool,
}


impl Word<Vec<char>> {
    pub fn new_owned(source: Vec<char>) -> Word<Vec<char>> {
        let len = source.len();
        let chars = source.clone();
        Word {
            source,
            slice: (0, len),
            chars,
            fin: true,
         }
    }
}


impl<'a> Word<Cow<'a, [char]>> {
    pub fn new_cow(source: Cow<'a, [char]>) -> Word<Cow<'a, [char]>> {
        let len = source.len();
        let chars = match source {
            Cow::Borrowed(s) => Cow::Borrowed(&s[..]),
            Cow::Owned(ref s) => Cow::Owned(s[..].to_vec()),
        };
        Word {
            source,
            slice: (0, len),
            chars,
            fin: true,
         }
    }
}


impl<T: AsRef<[char]>> Word<T> {
    pub fn fin(mut self, fin: bool) -> Self {
        self.fin = fin;
        self
    }

    pub fn len(&self) -> usize {
        self.chars.as_ref().len()
    }

    pub fn is_empty(&self) -> bool {
        self.chars.as_ref().is_empty()
    }

    pub fn to_owned(&self) -> Word<Vec<char>> {
        Word {
            source: self.source.as_ref().to_vec(),
            slice:  self.slice,
            chars:  self.chars.as_ref().to_vec(),
            fin:    self.fin,
        }
    }

    pub fn to_cow<'a>(&'a self) -> Word<Cow<'a, [char]>> {
        Word {
            source: Cow::Borrowed(self.source.as_ref()),
            slice:  self.slice,
            chars:  Cow::Borrowed(self.chars.as_ref()),
            fin:    self.fin,
        }
    }
}


impl<T: AsRef<[char]>> fmt::Debug for Word<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Word {{ \"")?;
        for ch in self.chars.as_ref().iter() {
            write!(f, "{}", ch)?;
        }
        write!(f, "\"")?;
        if !self.fin { write!(f, "..")?; }
        write!(f, " }}")?;
        Ok(())
    }
}
