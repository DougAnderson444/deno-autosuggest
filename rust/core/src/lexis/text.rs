use std::fmt;
use super::CharPattern;
use super::Word;


#[derive(PartialEq)]
pub struct Text<T: AsRef<[char]>> {
    pub source: T,
    pub words: Vec<Word<T>>,
}


impl Text<Vec<char>> {
    pub fn from_vec(source: Vec<char>) -> Text<Vec<char>> {
        let cloned = source.clone();
        Text {
            source: source,
            words: vec![Word::from_vec(cloned)],
        }
    }

    pub fn from_str(source: &str) -> Text<Vec<char>> {
        let source = source.chars().collect::<Vec<_>>();
        let word   = source.clone();
        Text {
            source,
            words: vec![Word::from_vec(word)],
        }
    }

    pub fn to_ref<'a>(&'a self) -> Text<&'a [char]> {
        Text {
            source: &self.source,
            words: self.words.iter().map(|w| w.to_ref()).collect()
        }
    }

    pub fn split<P: CharPattern>(mut self, pattern: &P) -> Self {
        let mut words = Vec::with_capacity(self.words.len());
        for word in &self.words {
            for splitted in word.split(pattern) {
                words.push(splitted.to_own());
            }
        }
        self.words = words;
        self
    }

    pub fn strip<P: CharPattern>(mut self, pattern: &P) -> Self {
        for word in &mut self.words {
            word.strip(pattern);
        }
        self.words.retain(|w| w.len() > 0);
        self
    }

    pub fn lower(mut self) -> Self {
        for word in &mut self.words {
            word.lower();
        }
        self
    }
}


impl<'a> Text<&'a [char]> {
    pub fn to_own(&self) -> Text<Vec<char>> {
        Text {
            source: self.source.to_vec(),
            words:  self.words.iter().map(|w| w.to_own()).collect(),
        }
    }
}


impl<T: AsRef<[char]>> Text<T> {
    pub fn fin(mut self, fin: bool) -> Self {
        if let Some(word) = self.words.last_mut() {
            word.fin = fin;
        }
        self
    }

    pub fn is_empty(&self) -> bool {
        self.words.is_empty()
    }
}


impl<T: AsRef<[char]>> fmt::Debug for Text<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Text {{")?;
        for word in &self.words {
            write!(f, " \"")?;
            for ch in word.chars.as_ref().iter() {
                write!(f, "{}", ch)?;
            }
            write!(f, "\"")?;
            if !word.fin { write!(f, "..")?; }
        }
        write!(f, " }}")?;
        Ok(())
    }
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use super::{Word, Text};
    use super::super::Chars;

    use Chars::{
        Whitespaces,
        Punctuation,
    };

    #[test]
    fn text_split() {
        let t = Text::from_str(" Foo Bar, Baz; ").split(&[Whitespaces, Punctuation]);
        assert_debug_snapshot!(t);
    }

    #[test]
    fn text_split_empty() {
        let t = Text::from_str(", ").split(&[Whitespaces, Punctuation]);
        assert_debug_snapshot!(t);
    }

    #[test]
    fn text_split_unfinished() {
        let t1 = Text::from_str(" Foo Bar, Baz"  ).fin(false).split(&[Whitespaces, Punctuation]);
        let t2 = Text::from_str(" Foo Bar, Baz; ").fin(false).split(&[Whitespaces, Punctuation]);
        assert_eq!(t1.words.last().unwrap().fin, false);
        assert_eq!(t2.words.last().unwrap().fin, true);
    }

    #[test]
    fn text_strip() {
        let t = Text {
                source: vec![],
                words: vec![
                    Word::from_str("-Foo-"),
                    Word::from_str(","),
                    Word::from_str("Baz; "),
                ],
            }
            .strip(&[Whitespaces, Punctuation]);
        assert_debug_snapshot!(t);
    }

    #[test]
    fn text_strip_unfinished() {
        let t1 = Text {
                source: vec![],
                words: vec![
                    Word::from_str("-Foo-"),
                    Word::from_str("Baz"),
                ],
            }
            .fin(false)
            .strip(&[Whitespaces, Punctuation]);
        let t2 = Text {
                source: vec![],
                words: vec![
                    Word::from_str("-Foo-"),
                    Word::from_str("Baz; "),
                ],
            }
            .fin(false)
            .strip(&[Whitespaces, Punctuation]);
        assert_eq!(t1.words.last().unwrap().fin, false);
        assert_eq!(t2.words.last().unwrap().fin, true);
    }

    #[test]
    fn text_lower() {
        let t  = Text {
                source: vec![],
                words: vec![
                    Word::from_str("Foo,"),
                    Word::from_str("Bar"),
                    Word::from_str("Baz"),
                ],
            }.lower();
        assert_debug_snapshot!(t);
    }
}
