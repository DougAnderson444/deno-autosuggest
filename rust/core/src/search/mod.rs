mod score;
mod highlight;

use std::borrow::Cow;
use std::cmp::Ordering;
use crate::lexis::{Text, Chars};

pub use score::{score, Scores};
pub use highlight::highlight;


#[derive(Clone, Debug)]
pub struct Record {
    id: usize,
    text: Text<Vec<char>>,
}


impl Record {
    pub fn new(id: usize, source: &[char]) -> Record {
        Record {
            id,
            text: tokenize_record(source).to_owned(),
        }
    }

    pub fn to_hit<'a>(&'a self) -> Hit<Cow<'a, [char]>> {
        Hit {
            id: self.id,
            text: self.text.to_cow(),
            scores: Default::default(),
        }
    }
}


#[derive(Debug, Clone)]
pub struct Hit<T: AsRef<[char]>> {
    pub id:      usize,
    pub text:    Text<T>,
    pub scores:  Scores,
}


impl<T: AsRef<[char]>> Hit<T> {
    pub fn new(id: usize, text: Text<T>) -> Hit<T> {
        Hit { id, text, scores: Default::default() }
    }
}


pub fn search<T: AsRef<[char]>, U: AsRef<[char]>>(query: &Text<T>, hits: &mut Vec<Hit<U>>) {
    score(query, hits);
    filter(query, hits);
    sort(hits);
}


pub fn filter<T: AsRef<[char]>, U: AsRef<[char]>>(query: &Text<T>, hits: &mut Vec<Hit<U>>) {
    if query.is_empty() {
        return;
    }
    hits.retain(|hit| {
        let matches = &hit.scores.matches;
        if matches.len() == 0 { return false; }
        if matches.len() == 1 && query.words.len() > 1 {
            let word_match = &matches[0];
            let unfinished = !word_match.fin;
            let first_half = (word_match.query.len * 2) < word_match.record.len;
            if unfinished && first_half { return false; }
        }
        true
    });
}


pub fn sort<T: AsRef<[char]>>(hits: &mut Vec<Hit<T>>) {
    hits.sort_by(|hit1, hit2| {
        let s1 = &hit1.scores;
        let s2 = &hit2.scores;
        if s1.matches.len() != s2.matches.len() { return s1.matches.len().cmp(&s2.matches.len()).reverse(); }
        if s1.typos  != s2.typos  { return s1.typos.cmp(&s2.typos); }
        if s1.trans  != s2.trans  { return s1.trans.cmp(&s2.trans); }
        if s1.fin    != s2.fin    { return s1.fin.cmp(&s2.fin).reverse(); }
        if s1.offset != s2.offset { return s1.offset.cmp(&s2.offset); }
        Ordering::Equal
    });
}


pub fn tokenize_record<'a>(source: &'a [char]) -> Text<Cow<'a, [char]>> {
    Text::new_cow(Cow::Borrowed(source))
        .split(&[Chars::Whitespaces, Chars::Control])
        .strip(&[Chars::NotAlphaNum])
        .lower()
}


pub fn tokenize_query<'a>(source: &'a [char]) -> Text<Cow<'a, [char]>> {
    Text::new_cow(Cow::Borrowed(source))
        .fin(false)
        .split(&[Chars::Whitespaces, Chars::Control, Chars::Punctuation])
        .strip(&[Chars::NotAlphaNum])
        .lower()
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::lexis::{Text, Chars};
    use super::{Hit, search};
    use std::borrow::Cow;

    fn chars(s: &str) -> Vec<char> {
        s.chars().collect()
    }

    fn record<'a>(chars: &'a [char]) -> Text<Cow<'a, [char]>> {
        Text::new_cow(Cow::Borrowed(&chars[..]))
            .split(&Chars::Whitespaces)
    }

    fn query<'a>(s: &'a [char]) -> Text<Cow<'a, [char]>> {
        record(s).fin(false)
    }

    #[test]
    fn search_empty() {
        let cr1 = chars("brown plush bear");
        let cr2 = chars("metal detector");
        let cr3 = chars("yellow metal mailbox");

        let mut hits = vec![
            Hit::new(10, record(&cr1)),
            Hit::new(20, record(&cr2)),
            Hit::new(30, record(&cr3)),
        ];

        let cq = chars("");
        let q  = query(&cq);

        search(&q, &mut hits);

        assert_debug_snapshot!(hits);
    }

    #[test]
    fn search_equal() {
        let cr1 = chars("brown plush bear");
        let cr2 = chars("metal detector");
        let cr3 = chars("yellow metal mailbox");

        let mut hits = vec![
            Hit::new(10, record(&cr1)),
            Hit::new(20, record(&cr2)),
            Hit::new(30, record(&cr3)),
        ];

        let cq = chars("yelow metall maiblox");
        let q  = query(&cq);

        search(&q, &mut hits);

        assert_debug_snapshot!(hits);
    }

    #[test]
    fn search_partial() {
        let cr1 = chars("brown plush bear");
        let cr2 = chars("metal detector");
        let cr3 = chars("yellow metal mailbox");

        let mut hits1 = vec![
            Hit::new(10, record(&cr1)),
            Hit::new(20, record(&cr2)),
            Hit::new(30, record(&cr3)),
        ];
        let mut hits2 = hits1.clone();
        let mut hits3 = hits1.clone();

        let cq1 = chars("metall mailbox");
        let cq2 = chars("metall mail");
        let cq3 = chars("met");
        let q1  = query(&cq1);
        let q2  = query(&cq2);
        let q3  = query(&cq3);

        search(&q1, &mut hits1);
        search(&q2, &mut hits2);
        search(&q3, &mut hits3);

        assert_debug_snapshot!(hits1);
        assert_debug_snapshot!(hits2);
        assert_debug_snapshot!(hits3);
    }

    #[test]
    fn search_intersection() {
        let cr1 = chars("brown plush bear");
        let cr2 = chars("metal detector");
        let cr3 = chars("yellow metal mailbox");

        let mut hits1 = vec![
            Hit::new(10, record(&cr1)),
            Hit::new(20, record(&cr2)),
            Hit::new(30, record(&cr3)),
        ];
        let mut hits2 = hits1.clone();

        let cq1 = chars("red wooden mailbox");
        let cq2 = chars("red wooden mail");
        let q1  = query(&cq1);
        let q2  = query(&cq2);

        search(&q1, &mut hits1);
        search(&q2, &mut hits2);

        assert_debug_snapshot!(hits1);
        assert_debug_snapshot!(hits2);
    }

    #[test]
    fn search_min_match() {
        let cr1 = chars("brown plush bear");
        let cr2 = chars("metal detector");
        let cr3 = chars("yellow metal mailbox");

        let mut hits1 = vec![
            Hit::new(10, record(&cr1)),
            Hit::new(20, record(&cr2)),
            Hit::new(30, record(&cr3)),
        ];
        let mut hits2 = hits1.clone();

        let cq1 = chars("wooden mai");
        let cq2 = chars("wooden mail");
        let q1  = query(&cq1);
        let q2  = query(&cq2);

        search(&q1, &mut hits1);
        search(&q2, &mut hits2);

        assert_debug_snapshot!(hits1);
        assert_debug_snapshot!(hits2);
    }

    #[test]
    fn search_transpositions() {
        let cr1 = chars("metal mailbox");
        let cr2 = chars("mailbox metal");

        let mut hits = vec![
            Hit::new(10, record(&cr1)),
            Hit::new(20, record(&cr2)),
        ];

        let cq = chars("mailbox met");
        let q  = query(&cq);

        search(&q, &mut hits);

        assert_debug_snapshot!(hits);
    }
}
