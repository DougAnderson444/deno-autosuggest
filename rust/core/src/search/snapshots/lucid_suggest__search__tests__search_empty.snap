---
source: src/search/mod.rs
expression: hits
---
[
    Hit {
        id: 10,
        text: Text { "brown" "plush" "bear" },
        scores: Scores {
            matches: [],
            typos: 0,
            offset: 0,
            trans: 0,
            fin: true,
        },
    },
    Hit {
        id: 20,
        text: Text { "metal" "detector" },
        scores: Scores {
            matches: [],
            typos: 0,
            offset: 0,
            trans: 0,
            fin: true,
        },
    },
    Hit {
        id: 30,
        text: Text { "yellow" "metal" "mailbox" },
        scores: Scores {
            matches: [],
            typos: 0,
            offset: 0,
            trans: 0,
            fin: true,
        },
    },
]
